task-context.md

```diff
@@ -0,0 +1,193 @@
+- Feature Name: task-context
+- Start Date: 2018-01-22
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Learnability, reasoning, debuggability and `no_std`-compatibility improvements achieved by
+making task wakeup handles into explicit arguments.
+
+# Motivation
+[motivation]: #motivation
+
+## Learnability
+
+The `futures` task wakeup mechanism is [a regular source of confusion for new
+users][explicit task issue]. Looking at the signature of `Future::poll`, it's
+hard to tell how `Future`s work:
+
+```rust
+fn poll(&mut self) -> Poll<Self::Item, Self::Error>;
+```
+
+`poll` takes no arguments, and returns either `Ready`, `NotReady`, or an error.
+From the signature, it looks as though the only way to complete a `Future` is
+to continually call `poll` in a loop until it returns `Ready`. This is
+incorrect.
+
+Furthermore, when implementing `Future`, it's necessary to schedule a wakeup
+when you return `NotReady`, or else you might never be `poll`ed again. The
+trick to the API is that there is an implicit `Task` argument to `Future::poll`
+which is passed using thread-local storage. This `Task` argument must be
+`notify`d by the `Future` in order for the task to awaken and `poll` again.
+It's essential to remember to do this, and to ensure that the right schedulings
+occur on every code path.
+
+[explicit task issue]: https://github.com/alexcrichton/futures-rs/issues/129
+
+## Reasoning about code
+
+The implicit `Task` argument makes it difficult to tell which functions
+aside from `poll` will use the `Task` handle to schedule a wakeup. That has a
+few implications.
+
+First, it's easy to accidentally call a function that will attempt to access
+the task outside of a task context. Doing so will result in a panic, but it 
+would be better to detect this mistake statically.
+
+Second, and relatedly, it is hard to audit a piece of code for which calls
+might involve scheduling a wakeup--something that is critical to get right
+in order to avoid "lost wakeups", which are far harder to debug than an
+explicit panic.
+
+## Debuggability
+
+Today's implicit TLS task system provides us with fewer "hooks" for adding debugging
+tools to the task system. Deadlocks and lost wakeups are some of the trickiest things
+to track down with futures code today, and it's possible that by using a "two stage"
+system like the one proposed in this RFC, we will have more room for adding debugging 
+hooks to track the precise points at which queing for wakeup occurred, and so on.
```

Comment:
I don't see how this is true. Making the task argument explicit has no functionality change. You can hook into the task system to add debugging logic today. It is just that nobody has actually done it.